crates/ra_syntax/src/parser_impl.rs

```diff
@@ -178,6 +192,18 @@ impl<'t> ParserImpl<'t> {
         }
     }
 
+    /// Given events:
+    /// ```
+    /// [EVENT0, EVENT1, EVENT2, EVENT3]
+    ///          ^^^^^^          ^^^^^^= current_event
+    ///          |-- parent
+    /// ```
+    /// Suppose to precede 2 positions, save the position of event3's parent(`EVENT1`).
+    /// ```
+    /// EVENT3 {
+    ///     forward_parent = 1
+    /// }
+    /// ```
```

Comment:
This is backwards I think. If we call precede at `EVENT3`, it's the `EVENT1` forward-parent that get's changed. 

That is, normally `EVENT1` is a parent of `EVENT3`, but `precede` makes it so that `EVENT3` becomes the parrent of `EVENT1`. To do this, we need to add a mark on `EVENT1` that its parent is to be found down the line.